CIP-0132/README.md

```diff
@@ -0,0 +1,117 @@
+---
+CIP: 132
+Title: New Plutus Builtin dropList
+Status: Proposed
+Category: Plutus
+Authors:
+    - Philip DiSarro <info@anastasialabs.com>
+Implementors: []
+Discussions:
+    - https://github.com/cardano-foundation/CIPs/pull/767
+Created: 2024-02-25
+License: CC-BY-4.0
+---
+
+## Abstract
+This document describes the addition of a new Plutus builtin `dropList` with the signature `Integer -> List a -> List a` that drops a given number of elements the list. This drastically increases the efficiency of `elemAt` which is currently a huge throughput bottleneck for many DApps. 
+
+## Motivation: why is this CIP necessary?
+The deterministic script evaluation property of the ledger (also stated as "script interpreter arguments are fixed") is a unique characteristic of the Cardano ledger that allows us to perform powerful optimizations that are not possible in systems with indeterminstic script evaluation. For instance, searching for elements in a data structure can 
+be done entirely off-chain, and then we simply provide the onchain code with the index (via redeemer) to where the element we want to find is supposed to be, and then check (onchain) that it is indeed the element we were expecting. This design pattern of passing the indices of elements required for validation logic in the redeemer is commonly referred to as redeemer-indexing. 
+Even though it is still a very powerful optimization in its current state, it is currently bottlenecked by the lack of a builtin that applies tail a given number of times to a list. Currently, any implementation of `elemAt :: Integer -> List a -> a` or `drop` requires the use of the fixed point combinator (Y combinator) which has a significant cost in onchain code.
+                            
+Consider the naive approach:
+```haskell
+{- | Fixpoint recursion. Used to encode recursive functions.
+     Hopefully this illustrates the overhead that this incurs. 
+-}
+pfix :: Term s (((a :--> b) :--> a :--> b) :--> a :--> b)
+pfix = phoistAcyclic $
+  punsafeCoerce $
+    plam' $ \f ->
+      plam' (\(x :: Term s POpaque) -> f # plam' (\(v :: Term s POpaque) -> punsafeCoerce x # x # v))
+        # punsafeCoerce (plam' $ \(x :: Term s POpaque) -> f # plam' (\(v :: Term s POpaque) -> punsafeCoerce x # x # v))
+
+-- | Lazy if-then-else
+-- Two forces + two delays + builtinIfThenElse
+pif :: Term s PBool -> Term s a -> Term s a -> Term s a
+pif b case_true case_false = pforce $ (pforce $ punsafeBuiltin PLC.IfThenElse) # b # pdelay (f PTrue) # pdelay (f PFalse) 
+ where
+   f = \case
+     PTrue -> case_true
+     PFalse -> case_false
```

Comment:
Why not just inline `case_true` and `case_false` instead of using `f PTrue` and `f PFalse`?

Reply:
In Plutarch you can think of Haskell level work as a sort of pre-compilation step. The above would actually be inlined when compiled, because `f` or `\case` (a Haskell level function) is used on `PTrue` (a Haskell level argument), this would become:
`pif b case_true case_false = pforce $ (pforce $ punsafeBuiltin PLC.IfThenElse) # b # pdelay case_true # pdelay case_false`

after the pre-compilation step. I have edited it for clarity though.